common/src/message.rs

```diff
@@ -0,0 +1,488 @@
+                    }),
+                    OpCode::CreateTree => Some(Message::CreateTreeMessage {
+                        ptr: left,
+                        size: right,
+                    }),
+                    _ => unreachable!(),
+                }
+            }
+            _ => {
+                let left: &[u8; 9] = buf[0..9].try_into().expect("split slice with wrong length");
+                let left = ArcaHandle::parse(left);
+
+                match op {
+                    OpCode::CreateThunk => {
+                        if let ArcaHandle::BlobHandle(b) = left {
+                            Some(Message::CreateThunkMessage { handle: b })
+                        } else {
+                            None
+                        }
+                    }
+                    OpCode::RunThunk => {
+                        if let ArcaHandle::ThunkHandle(t) = left {
+                            Some(Message::RunThunkMessage { handle: t })
+                        } else {
+                            None
+                        }
+                    }
+                    OpCode::Apply => {
+                        if let ArcaHandle::LambdaHandle(l) = left {
+                            let right: &[u8; 9] = buf[9..18]
+                                .try_into()
+                                .expect("split slice with wrong length");
+                            let right = ArcaHandle::parse(right);
+                            Some(Message::ApplyMessage {
+                                lambda_handle: l,
+                                arg_handle: right,
+                            })
+                        } else {
+                            None
+                        }
+                    }
+                    OpCode::Reply => Some(Message::ReplyMessage { handle: left }),
+                    OpCode::Drop => Some(Message::DropMessage { handle: left }),
+                    _ => unreachable!(),
+                }
+            }
+        }
+    }
+
+    fn msg_size(op: &OpCode) -> usize {
+        match op {
+            OpCode::CreateBlob => 16,
+            OpCode::CreateTree => 16,
+            OpCode::CreateThunk => 9,
+            OpCode::RunThunk => 9,
+            OpCode::Apply => 18,
+            OpCode::Reply => 9,
+            OpCode::Drop => 9,
+        }
+    }
+}
+
+struct MessageParser<'a> {
+    pending_opcode: Option<OpCode>,
+    pending_opcode_buf: [u8; 1],
+    pending_payload: [u8; 18],
+    pending_payload_offset: usize,
+    expected_payload_size: usize,
+    rb: RingBufferReceiver<'a>,
+}
+
+impl<'a> MessageParser<'a> {
+    fn new(rb: RingBufferReceiver<'a>) -> MessageParser<'a> {
+        Self {
+            pending_opcode: None,
+            pending_opcode_buf: [0; 1],
+            pending_payload: [0; 18],
+            pending_payload_offset: 0,
+            expected_payload_size: 0,
+            rb,
+        }
+    }
+
+    // Read exactly once
```

Comment:
Is this comment accurate?

Reply:
This function only reads from the ring buffer once and may not return a message even if the whole message is already available in the ring buffer, so caller of this function needs to make call until the result is `Err` to drain the buffer. I think such behavior is correct given `std::io::Read`'s description, but I am not sure whether it is common for actual implementations of `read()`